src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadBufferState.cs

```diff
@@ -2,32 +2,168 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
 
 namespace System.Text.Json.Serialization
 {
     internal struct ReadBufferState : IDisposable
     {
-        public byte[] Buffer;
-        public int BytesInBuffer;
-        public int ClearMax;
-        public bool IsFirstIteration;
-        public bool IsFinalBlock;
+        private byte[] _buffer;
+        private byte _offset; // Read bytes offset typically used when skipping the UTF-8 BOM.
+        private int _count; // Number of read bytes yet to be consumed by the serializer.
+        private int _maxCount; // Number of bytes we need to clear before returning the buffer.
+        private bool _isFirstBlock;
+        private bool _isFinalBlock;
 
-        public ReadBufferState(int defaultBufferSize)
+        public ReadBufferState(int initialBufferSize)
         {
-            Buffer = ArrayPool<byte>.Shared.Rent(Math.Max(defaultBufferSize, JsonConstants.Utf8Bom.Length));
-            BytesInBuffer = ClearMax = 0;
-            IsFirstIteration = true;
-            IsFinalBlock = false;
+            _buffer = ArrayPool<byte>.Shared.Rent(Math.Max(initialBufferSize, JsonConstants.Utf8Bom.Length));
+            _maxCount = _count = _offset = 0;
+            _isFirstBlock = true;
+            _isFinalBlock = false;
+        }
+
+        public bool IsFinalBlock => _isFinalBlock;
+
+        public ReadOnlySpan<byte> Bytes => _buffer.AsSpan(_offset, _count);
+
+        /// <summary>
+        /// Read from the stream until either our buffer is filled or we hit EOF.
+        /// Calling ReadCore is relatively expensive, so we minimize the number of times
+        /// we need to call it.
+        /// </summary>
+        public readonly async ValueTask<ReadBufferState> ReadFromStreamAsync(
+            Stream utf8Json,
+            CancellationToken cancellationToken,
+            bool fillBuffer = true)
+        {
+            // Since mutable structs don't work well with async state machines,
+            // make all updates on a copy which is returned once complete.
+            ReadBufferState bufferState = this;
+
+            do
+            {
+                int bytesRead = await utf8Json.ReadAsync(
+#if BUILDING_INBOX_LIBRARY
+                    bufferState._buffer.AsMemory(bufferState._count),
+#else
+                    bufferState._buffer, bufferState._count, bufferState._buffer.Length - bufferState._count,
```

Comment:
isn't it better to keep count/length in the locals in the loop and then assign them back once you're done?

Reply:
Ultimately this will compile into an async state machine so any locals will be represented using fields in the SM struct.